src/utils/callState.ts

```diff
@@ -1,7 +1,40 @@
 import type { FunctionFragment, Interface } from '@ethersproject/abi'
+import deepEqual from 'fast-deep-equal'
+import { useMemo, useRef } from 'react'
 import { INVALID_CALL_STATE, LOADING_CALL_STATE } from '../constants'
 import type { CallResult, CallState, CallStateResult } from '../types'
 
+// Converts CallResult[] to CallState[], carrying over any decoded objects between calls.
+// This ensures that the CallState results remain referentially stable when unchanged, preventing
+// spurious re-renders which would otherwise occur because mapping always creates a new object.
+export function useStableCallStates(
+  results: CallResult[],
+  contractInterface: Interface | undefined,
+  fragment: ((i: number) => FunctionFragment | undefined) | FunctionFragment | undefined,
+  latestBlockNumber: number | undefined
+): CallState[] {
+  const lastStates = useRef<CallState[]>([])
+  return useMemo(() => {
+    const callStates = results.map((result, i) => {
+      const resultFragment = typeof fragment === 'function' ? fragment(i) : fragment
+      const callState = toCallState(result, contractInterface, resultFragment, latestBlockNumber)
+
+      // if the result has not changed, return a referentially stable result
+      const lastState = lastStates.current[i]
+      if (deepEqual(callState, lastState)) {
+        return lastState
+      }
+
+      return callState
+    })
+    // if no results have changed, return a referentially stable array
+    if (callStates.every((result, i) => result === lastStates.current[i])) return lastStates.current
```

Comment:
can you make this check in the results.map above, to reduce time complexity? 

you can set a boolean before the map on line 18 like "allCallStatesUnchanged = true" in the beginning, and set it to false upon the first instance where` result !== lastStates.current[i]`

Reply:
this would help us avoid the need for deepEqual (I think)